### PR TITLE
Huge refactor

### PR DESCRIPTION
**Pull request by gacikjakub**

### CODE REVIEW

**Comment by gacikjakub** [comment]
Build 46 Successful
Deployed under: shooting-refactor.battleships.kuba005.nazwa.pl

File: backend/src/main/java/pl/krkteam/battleships/authentication/AddPlayerResultDTO.java [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package pl.krkteam.battleships.authentication;
+
+public class AddPlayerResultDTO {
```
Comment (boczkas) [inline backend/src/main/java/pl/krkteam/battleships/authentication/AddPlayerResultDTO.java]: PlayerResultAdderDTO

File: backend/src/main/java/pl/krkteam/battleships/authentication/RegistrationController.java [hunk 2]
```diff
@@ -1,33 +1,38 @@
...
 
 @CrossOrigin(origins = "http://localhost:8080")
 @RestController
 public class RegistrationController {
 
-    PlayerHolder playerHolder = new PlayerHolder();
+    private final Game game;
+    private final PlayerHolder playerHolder;
+
+
```
Comment (boczkas) [inline backend/src/main/java/pl/krkteam/battleships/authentication/RegistrationController.java]: remove \n

File: backend/src/main/java/pl/krkteam/battleships/common/domain/Game.java [hunk 3]
```diff
@@ -1,11 +1,38 @@
 package pl.krkteam.battleships.common.domain;
 
-import pl.krkteam.battleships.opponent_shot_result_keeping.ShotResultQueueHolder;
+import org.springframework.stereotype.Component;
+import pl.krkteam.battleships.common.domain.player.Player;
+import pl.krkteam.battleships.opponent.shot.response.domain.ShotResultQueue;
+import pl.krkteam.battleships.opponent.shot.response.domain.ShotResultQueueHolder;
+import pl.krkteam.battleships.opponent.shot.response.dto.OpponentShotResult;
 
+@Component
```
Comment (boczkas) [inline backend/src/main/java/pl/krkteam/battleships/common/domain/Game.java:9]: self comment:
prototype class.

File: backend/src/main/java/pl/krkteam/battleships/common/domain/GameBoard.java [hunk 4]
```diff
@@ -22,14 +20,20 @@ public ShipHolder getShipHolder() {
     }
 
     public Ship createShip(List<Coordinates> coordinates) {
-        Ship ship = new Ship();
-        coordinates.forEach(coor -> {
-            PresentMast presentMast = new PresentMast();
-            ship.addMast(presentMast);
-            board.putCoordinatesAndCell(coor, presentMast);
+        Ship ship = new Ship(shipHolder);
+        pupMastsOnBoard(coordinates, ship);
```
Comment (boczkas) [inline backend/src/main/java/pl/krkteam/battleships/common/domain/GameBoard.java]: put maybe? ;)

File: backend/src/main/java/pl/krkteam/battleships/common/domain/GameBoardHolder.java [hunk 5]
```diff
@@ -33,4 +31,13 @@ public GameBoard getGameBoard(Player player) {
         }
         return playerGameBoardMap.get(player);
     }
+
+    public Player getOpponentPlayer(Player player) {
+        Set<Player> playerSet = playerGameBoardMap.keySet();
+        List<Player> playerList = new ArrayList<>(playerSet);
+        playerList.remove(player);
```
Comment (boczkas) [inline backend/src/main/java/pl/krkteam/battleships/common/domain/GameBoardHolder.java:38]: suggestion:
playerList.indexOf(player) + 1) % 2)

File: backend/src/main/java/pl/krkteam/battleships/common/domain/ship/Ship.java [hunk 6]
```diff
@@ -1,22 +1,46 @@
...
-    public void addMast(final Mast mast) {
+    public void attachMast(final Mast mast) {
         mastList.add(mast);
     }
 
     public List<Mast> getMastList() {
-        return Collections.unmodifiableList(mastList);
+        return mastList;
+    }
+
```
Comment (boczkas) [inline backend/src/main/java/pl/krkteam/battleships/common/domain/ship/Ship.java]: \n remove

File: backend/src/main/java/pl/krkteam/battleships/common/domain/ship/Ship.java [hunk 7]
```diff
@@ -1,22 +1,46 @@
...
+    }
 
-    public void addMast(final Mast mast) {
+    public void attachMast(final Mast mast) {
         mastList.add(mast);
     }
 
     public List<Mast> getMastList() {
-        return Collections.unmodifiableList(mastList);
+        return mastList;
```
Comment (boczkas) [inline backend/src/main/java/pl/krkteam/battleships/common/domain/ship/Ship.java]: unmodifiable list

File: backend/src/main/java/pl/krkteam/battleships/common/domain/ship/Ship.java [hunk 8]
```diff
@@ -1,22 +1,46 @@
...
+    }
+
+
+    public void notifyMastChangeState() {
+        final Optional<Mast> anyPresentMast = mastList.stream().filter(
+                mast -> mast.getMastState().equals(MastState.MAST_PRESENT)).findAny();
```
Comment (boczkas) [inline backend/src/main/java/pl/krkteam/battleships/common/domain/ship/Ship.java]: findAny() to new line

File: backend/src/main/java/pl/krkteam/battleships/common/domain/ship/Ship.java [hunk 9]
```diff
@@ -1,22 +1,46 @@
...
+            shipHolder.notifyShipChangeState();
+        }
+    }
+
+    public ShipState getShipState() {
+        return shipState;
+    }
+
+    public ShipHolder getShipHolder() {
+        return shipHolder;
```
Comment (boczkas) [inline backend/src/main/java/pl/krkteam/battleships/common/domain/ship/Ship.java:47]: consider unmodifiable shipHolder

File: backend/src/main/java/pl/krkteam/battleships/common/domain/ship/ShipHolder.java [hunk 10]
```diff
@@ -2,15 +2,29 @@
...
+    void attachShip(final Ship ship) {
         shipList.add(ship);
     }
 
     public Ship getShip(int index) {
         return shipList.get(index);
     }
+
+    void notifyShipChangeState() {
+        final Optional<Ship> anyFloatingShip = shipList.stream().filter(
```
Comment (boczkas) [inline backend/src/main/java/pl/krkteam/battleships/common/domain/ship/ShipHolder.java]: use formating for stream here

File: backend/src/main/java/pl/krkteam/battleships/opponent/shot/response/controllers/OpponentResultController.java [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+    }
+
+    @GetMapping(value = "/game/opponent/result", produces = MediaType.APPLICATION_JSON_VALUE)
+    public String onShotOpponentResponse(@RequestParam String playerName) {
+        OpponentShotResult shotResult = getResponse(playerName);
+
+        Gson gson = new Gson();
+        return gson.toJson(shotResult);
+    }
+
```
Comment (boczkas) [inline backend/src/main/java/pl/krkteam/battleships/opponent/shot/response/controllers/OpponentResultController.java:30]: \n remove

File: backend/src/main/java/pl/krkteam/battleships/opponent/shot/response/dto/OpponentShotPlayerLoose.java [hunk 12]
```diff
@@ -0,0 +1,7 @@
+package pl.krkteam.battleships.opponent.shot.response.dto;
+
+public class OpponentShotPlayerLoose extends OpponentShotResult {
```
Comment (boczkas) [inline backend/src/main/java/pl/krkteam/battleships/opponent/shot/response/dto/OpponentShotPlayerLoose.java:3]: OpponentActionResult?

File: backend/src/main/java/pl/krkteam/battleships/opponent/shot/response/dto/OpponentShotResult.java [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+    private String opponentShotResult;
+    private ShotDTO opponentShotDTO;
+
+    public String getMessage() {
+        return opponentShotResult;
+    }
+
+    public OpponentShotResult(String opponentShotResult, int rowIndex, int columnIndex) {
+        this.opponentShotResult = opponentShotResult;
+        final CoordinateDTO shotCoordinate = new CoordinateDTO(columnIndex, rowIndex);
```
Comment (boczkas) [inline backend/src/main/java/pl/krkteam/battleships/opponent/shot/response/dto/OpponentShotResult.java]: extract new from constructor

File: backend/src/main/java/pl/krkteam/battleships/ships/placing/validation/ShipController.java [hunk 14]
```diff
@@ -40,18 +42,37 @@ public String communicateWithAngularByPostingShip(@RequestBody String post) {
...
 
-        ShipHolderDTO shipHolderDTO = gson.fromJson(post, ShipHolderDTO.class);
-        final ShipHolderFromJson shipHolderFromJson = shipsToShipHolder.convert(shipHolderDTO);
+        final ShipHolderFromJson shipHolderFromJson = createShipHolderFromJson(shipsJson);
 
         final PlacingValidationResultDTO placingValidationResultDTO =
-                shipsLocationValidatorService.validateShipLocation(shipHolderFromJson, gameBoard);
+                shipsLocationValidatorService.validateShipLocation(shipHolderFromJson, playerGameBoard);
 
+        afterValidation(placingValidationResultDTO, playerGameBoard);
```
Comment (boczkas) [inline backend/src/main/java/pl/krkteam/battleships/ships/placing/validation/ShipController.java]: resetGameBoardIfValidationFailed

File: backend/src/main/java/pl/krkteam/battleships/ships/placing/validation/ShipController.java [hunk 15]
```diff
@@ -40,18 +42,37 @@ public String communicateWithAngularByPostingShip(@RequestBody String post) {
...
+
+    private ShipHolderFromJson createShipHolderFromJson(String postData) {
+        Gson gson = new Gson();
+        final ShipHolderDTO shipHolderDTO = gson.fromJson(postData, ShipHolderDTO.class);
+        return shipsToShipHolder.convert(shipHolderDTO);
+    }
+
+    private GameBoard prepareGameBoard(String playerName) {
+        Player player = new Player(playerName);
+        GameBoard playerGameBoard = game.getGameBoardHolder().getGameBoard(player);
```
Comment (boczkas) [inline backend/src/main/java/pl/krkteam/battleships/ships/placing/validation/ShipController.java]: 1 dot per line

File: backend/src/main/java/pl/krkteam/battleships/shooting/controller/ShootingController.java [hunk 16]
```diff
@@ -1,68 +1,68 @@
...
-      shotResultDTO = new ResultSunkDTO();
-      mockedBoard.hittedMastCountForFirstShip = 0;
+    private ShotResultDTO getShotResult(Player opponentPlayer, ShotDTO shotDTO) {
+        final GameBoardHolder gameBoardHolder = game.getGameBoardHolder();
+        final GameBoard opponentGameBoard = gameBoardHolder.getGameBoard(opponentPlayer);
+        return shotResultCheckerService.checkShotResult(shotDTO, opponentGameBoard);
     }
 
-    System.out.println("hittedMastCountForSecondShip" + mockedBoard.hittedMastCountForSecondShip);
+    private void setResponseToOpponentPlayer(ShotResultDTO shotResultDTO, ShotDTO shotDTO, Player opponentPlayer) {
```
Comment (boczkas) [inline backend/src/main/java/pl/krkteam/battleships/shooting/controller/ShootingController.java]: sendResponseToOpponentPlayer

File: backend/src/main/java/pl/krkteam/battleships/shooting/dto/ShotDTO.java [hunk 17]
```diff
@@ -0,0 +1,22 @@
...
+
+public class ShotDTO {
+
+    private CoordinateDTO shotCoordinate;
+
+    public ShotDTO(CoordinateDTO shotCoordinate) {
+        this.shotCoordinate = shotCoordinate;
+    }
+
+    @Override
```
Comment (boczkas) [inline backend/src/main/java/pl/krkteam/battleships/shooting/dto/ShotDTO.java]: toString at the end of class

File: backend/src/main/java/pl/krkteam/battleships/shooting/services/CheckHit.java [hunk 18]
```diff
@@ -0,0 +1,19 @@
+package pl.krkteam.battleships.shooting.services;
```
Comment (boczkas) [inline backend/src/main/java/pl/krkteam/battleships/shooting/services/CheckHit.java:1]: Like for chain of responsibility

File: backend/src/main/java/pl/krkteam/battleships/shooting/services/ShotResultCheckerServiceImpl.java [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+import pl.krkteam.battleships.shooting.conventers.CoordinateDTOtoCoordinates;
+import pl.krkteam.battleships.shooting.dto.ShotDTO;
+import pl.krkteam.battleships.shooting.dto.result.ResultWrongShotDTO;
+import pl.krkteam.battleships.shooting.dto.result.ShotResultDTO;
+
+@Service
+public class ShotResultCheckerServiceImpl implements ShotResultCheckerService {
+    private final CoordinateDTOtoCoordinates coordinateDTOtoCoordinates;
+
+
```
Comment (boczkas) [inline backend/src/main/java/pl/krkteam/battleships/shooting/services/ShotResultCheckerServiceImpl.java]: \n remove

File: backend/src/main/java/pl/krkteam/battleships/shooting/services/UnexpectedShotResultException.java [hunk 20]
```diff
@@ -0,0 +1,18 @@
+package pl.krkteam.battleships.shooting.services;
+
+import pl.krkteam.battleships.common.domain.Coordinates;
+
+public class UnexpectedShotResultException extends Exception {
+
+    private Coordinates shotCoor;
```
Comment (boczkas) [inline backend/src/main/java/pl/krkteam/battleships/shooting/services/UnexpectedShotResultException.java]: change shotCoor name

File: backend/src/test/java/pl/krkteam/battleships/common/domain/GameBoardTest.java [hunk 21]
```diff
@@ -13,9 +13,18 @@
 import static org.testng.Assert.assertTrue;
 
 public class GameBoardTest {
+
     @Test
     public void testGetBoard() {
```
Comment (boczkas) [inline backend/src/test/java/pl/krkteam/battleships/common/domain/GameBoardTest.java]: to general name

File: backend/src/test/java/pl/krkteam/battleships/shooting/services/CheckMissTest.java [hunk 22]
```diff
@@ -0,0 +1,58 @@
...
+
+        final Board board = gameBoard.getBoard();
+
+        Coordinates shotCoor=new Coordinates(4,4);
+
+        ChainOfShotResult chainOfShotResult=new CheckMiss();
+
+        assertNotEquals(chainOfShotResult.getShotResult(shotCoor,board).getClass(),ResultMissDTO.class);
+    }
+}
```
Comment (boczkas) [inline backend/src/test/java/pl/krkteam/battleships/shooting/services/CheckMissTest.java]: \n

File: backend/src/test/java/pl/krkteam/battleships/shooting/services/CheckPlayerWonTest.java [hunk 23]
```diff
@@ -0,0 +1,46 @@
...
+        assertEquals(chainOfShotResult.getShotResult(shotCoor, board).getClass(),ResultHitDTO.class);
+
+        shotCoor = new Coordinates(5, 4);
+        chainOfShotResult.getShotResult(shotCoor, board);
+        assertEquals(chainOfShotResult.getShotResult(shotCoor, board).getClass(),ResultHitDTO.class);
+
+        shotCoor = new Coordinates(6, 4);
+        assertEquals(chainOfShotResult.getShotResult(shotCoor, board).getClass(),ResultPlayerWonDTO.class);
+    }
+}
```
Comment (boczkas) [inline backend/src/test/java/pl/krkteam/battleships/shooting/services/CheckPlayerWonTest.java]: \n

File: backend/src/test/java/pl/krkteam/battleships/shooting/services/CheckSunkTest.java [hunk 24]
```diff
@@ -0,0 +1,74 @@
...
+
+
+        final Coordinates coordinatesB1 = new Coordinates(8, 4);
+        List<Coordinates> coordinatesListB = Arrays.asList(coordinatesB1);
+
+        GameBoard gameBoard = new GameBoard();
+        gameBoard.createShip(coordinatesListA);
+        gameBoard.createShip(coordinatesListB);
+
+
```
Comment (boczkas) [inline backend/src/test/java/pl/krkteam/battleships/shooting/services/CheckSunkTest.java:33]: remove \n

File: backend/src/test/java/pl/krkteam/battleships/shooting/services/CheckSunkTest.java [hunk 25]
```diff
@@ -0,0 +1,74 @@
...
+        final Board board = gameBoard.getBoard();
+        ChainOfShotResult chainOfShotResult = new CheckMiss();
+
+        Coordinates shotCoor = new Coordinates(8, 4);
+        chainOfShotResult.getShotResult(shotCoor, board);
+
+        assertEquals(chainOfShotResult.getShotResult(shotCoor, board).getClass(), ResultSunkDTO.class);
+    }
+
+
```
Comment (boczkas) [inline backend/src/test/java/pl/krkteam/battleships/shooting/services/CheckSunkTest.java:43]: remove \n

File: backend/src/test/java/pl/krkteam/battleships/shooting/services/CheckSunkTest.java [hunk 26]
```diff
@@ -0,0 +1,74 @@
...
+        Coordinates shotCoor = new Coordinates(8, 4);
+        chainOfShotResult.getShotResult(shotCoor, board);
+
+        shotCoor = new Coordinates(8, 5);
+        chainOfShotResult.getShotResult(shotCoor, board);
+
+        assertEquals(chainOfShotResult.getShotResult(shotCoor, board).getClass(), ResultSunkDTO.class);
+    }
+}
```
Comment (boczkas) [inline backend/src/test/java/pl/krkteam/battleships/shooting/services/CheckSunkTest.java]: \n

File: backend/src/test/java/pl/krkteam/battleships/shooting/services/ShotResultCheckerServiceImplTest.java [hunk 27]
```diff
@@ -0,0 +1,44 @@
...
+        List<Coordinates> coordinatesListA = Arrays.asList(coordinates1, coordinates2,
+                coordinates3, coordinates4);
+
+        GameBoard gameBoard = new GameBoard();
+        gameBoard.createShip(coordinatesListA);
+
+        ShotDTO shotDTO = new ShotDTO(new CoordinateDTO(4, 3));
+        assertEquals(shotResultCheckerService.checkShotResult(shotDTO, gameBoard).getClass(), ResultHitDTO.class);
+    }
+}
```
Comment (boczkas) [inline backend/src/test/java/pl/krkteam/battleships/shooting/services/ShotResultCheckerServiceImplTest.java]: \n

**Review by boczkas — CHANGES_REQUESTED** [review]
cosmetic changes

**Review by JaroslawZaczyk — APPROVED** [review]
I hope Krzysiek is right and I can't tell anything about this code.

**Review by boczkas — APPROVED** [review]
:+1:

**Review by krzysieksr — APPROVED** [review]

**Pull request merged**